API tokens with expiration date

mawinter69 opened the pull request:
API tokens can now get an optional expiration date. Added a select drop down from where you can choose from predefined durations (default is 30 days), a custom date (max 1 year in the selector from the date input field) and no expiration. Expiration date is shown in the list. Not expiring tokens and tokens that expire in less than 7 days are marked with warning, expired tokens are marked with error.
The sidepanel on the user page now has a badge for the security action when there are tokens that expired or are about to expire. In the header the user avatar has a badge and in the dropdown is a badge.

<!-- Comment:
A great PR typically begins with the line below.
Replace <issue-number> with the issue number.
-->

Fixes #16695
<img width="1338" height="652" alt="image" src="https://github.com/user-attachments/assets/ca80138c-717b-42a5-a9ca-6f86d5f93971" />

<img width="631" height="566" alt="image" src="https://github.com/user-attachments/assets/7bf6e488-637c-486b-99c7-ea37e4d7efc9" />
<img width="557" height="735" alt="image" src="https://github.com/user-attachments/assets/239b302f-cc5f-4203-8847-2a0098ccf40b" />


<!-- Comment:
If the issue is not fully described in the issue tracker, add more information here (justification, pull request links, etc.).

 * We do not require an issue for minor improvements.
 * Major new features should have an issue created.
-->

### Testing done
Interactive testing for the UI part
Added tests for the scripting part, ensuring that expired tokens are not accepted

<!-- Comment:
Provide a clear description of how this change was tested.
At minimum this should include proof that a computer has executed the changed lines.
Ideally this should include an automated test or an explanation as to why this change has no tests.
Note that automated test coverage is less than complete, so a successful PR build does not necessarily imply that a computer has executed the changed lines.
If automated test coverage does not exist for the lines you are changing, you must describe the scenario(s) in which you manually tested the change.
For frontend changes, include screenshots of the relevant page(s) before and after the change.
For refactoring and code cleanup changes, exercise the code before and after the change and verify the behavior remains the same.
-->

### Proposed changelog entries

- API tokens with expiration date

<!-- Comment:
The changelog entry should be in the imperative mood; e.g., write "do this"/"return that" rather than "does this"/"returns that".
For examples, see: https://www.jenkins.io/changelog/

Do not include the issue in the changelog entry.
Include the issue in the description of the pull request so that the changelog generator can find it and include it in the generated changelog.

You may add multiple changelog entries if applicable by adding a new entry to the list, e.g.
- First changelog entry
- Second changelog entry
-->

### Proposed changelog category

/label rfe

<!--
The changelog entry needs to have a category which is selected based on the label.
If there's no changelog then the label should be `skip-changelog`.

The available categories are:
* bug - Minor bug. Will be listed after features
* developer - Changes which impact plugin developers
* dependencies - Pull requests that update a dependency
* internal - Internal only change, not user facing
* into-lts - Changes that are backported to the LTS baseline
* localization - Updates localization files
* major-bug - Major bug. Will be highlighted on the top of the changelog
* major-rfe - Major enhancement. Will be highlighted on the top
* rfe - Minor enhancement
* regression-fix - Fixes a regression in one of the previous Jenkins releases
* removed - Removes a feature or a public API
* skip-changelog - Should not be shown in the changelog

Non-changelog categories:
* web-ui - Changes in the web UI

Non-changelog categories require a changelog category but should be used if applicable,
comma separate to provide multiple categories in the label command.
-->

### Proposed upgrade guidelines

N/A

<!-- Comment:
Leave the proposed upgrade guidelines in the pull request with the "N/A" value if no upgrade guidelines are needed.
The changelog generator relies on the presence of the upgrade guidelines section as part of its data extraction process.
-->

### Submitter checklist

- [ ´x] The issue, if it exists, is well-described.
- [ ]xThe changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] UI changes do not introduce regressions when enforcing the current default rules of [Content Security Policy Plugin](https://plugins.jenkins.io/csp/). In particular, new or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@mention

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/core-pr-reviewers.
-->

Before the changes are marked as `ready-for-merge`:

### Maintainer checklist

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, be a _Bug_ or _Improvement_, and either the issue or pull request must be labeled as `lts-candidate` to be considered.

mawinter69 committed:
API tokens with expiration

API tokens can now get an optional expiration date.
Added a select drop down from where you can choose from predefined
durations, a custom date (max 1 year) and no expiration.
Expiration date is shown in the list. Not expiring tokens are marked
with warning.

fixes #16695

mawinter69 committed:
revert changes in SetupWizard

mawinter69 commented:
The SetupWizard allows to generate an API Token. Should that be automatically an expiring token maybe?

mawinter69 committed:
fix linter

mawinter69 committed:
fix ordering

timja commented:
> The SetupWizard allows to generate an API Token. Should that be automatically an expiring token maybe?

how do you generate one in the setup wizard?

timja commented on core/src/main/resources/jenkins/security/ApiTokenProperty/config.properties:
```diff
@@ -18,3 +18,4 @@ RevokedToken=The legacy token has been revoked
 NoCreationDate=There is no creation date for that token
 NoCreationDateValue=Unknown
 ConfirmRevokeSingleTitle=Revoke token
+tokenExpiresOn=Expires on {0}
```
What about when the token is expired? How does that look in the UI?

mawinter69 replied:
yep, some that is missing so far

lemeurherve commented on core/src/main/java/jenkins/security/ApiTokenProperty.java:
```diff
@@ -255,13 +256,20 @@ public static class TokenInfoAndStats {
             this.isLegacy = token.isLegacy();
+            LocalDate expirationDate = token.getExpirationDate();
+            if (expirationDate == null) {
+                this.expirationDate = "never";
```
That would be a breaking change maybe for later, but how about having a short-lived token generated by default instead of an eternal one to prevent mistakes, and to ensure eternal token creation is a conscencious and deliberate choice?

(EDIT: The line I put this comment on isn't the proper one, sorry)

mawinter69 replied:
In the UI the default is set to 30 days.

mawinter69 commented:
> > The SetupWizard allows to generate an API Token. Should that be automatically an expiring token maybe?
> 
> how do you generate one in the setup wizard?

You can set a system property then it generates an api token or uses a given one. See https://github.com/jenkinsci/jenkins/blob/46b6e75cf0bf873274c49af2534f84dceafefa02/core/src/main/java/jenkins/install/SetupWizard.java#L104

mawinter69 committed:
show when token expired

mawinter69 commented:
> What is default expiration ? Is it "never" or some sane time like 1 month ? What happens near expiration ? Is there some notification(on UI or mail) before , e.g. week before? It would be shame to expire without notificating user in advance

In the UI I set the default to 30 days. Notifications make sense. Maybe I can add some badge to the user icon in the header to inform about expiring tokens.
Sending mail notification is not possible out of the box as there is no support for mail sending in core. You need at least the mailer plugin

mawinter69 committed:
notify on expiring and expired tokens

mawinter69 commented:
Should expired tokens be automatically removed? Instantly, after a week, configurable or never?

mawinter69 committed:
Merge branch 'master' into expiring-tokens

timja commented:
> Should expired tokens be automatically removed? Instantly, after a week, configurable or never?

I would leave them, most systems I can think leave them around.

mawinter69 committed:
format

mawinter69 committed:
Merge branch 'master' into expiring-tokens

timja committed:
Merge branch 'master' into expiring-tokens

timja commented:
The balance of this feels a bit off:
<img width="742" height="396" alt="image" src="https://github.com/user-attachments/assets/622311a0-2170-473b-b6c2-a970df325873" />

Adding a third line really throws off alignment, especially with the buttons being on the middle line.
Do we need the created date at all? Isn't last used more important?

-----

Date format is inconsistent:

<img width="422" height="125" alt="image" src="https://github.com/user-attachments/assets/8fba3003-ef87-4fae-a0c9-e803ab5bf082" />

<img width="202" height="31" alt="image" src="https://github.com/user-attachments/assets/7d8bc6ca-860e-40e1-88c1-3432900e9316" />

Date picker appears to use a locale based format, where-as the two places the date is shown are using YYYY-MM-DD.
Ideally it would display in a locale specific format, I think there's a tag for that, `i:formatDate`?

-----

Apart from above looks great.

mawinter69 committed:
2 lines only

make the creation date a tooltip of the expiration

mawinter69 committed:
feedback

timja reviewed (approved):
Overall it LGTM.

I added a few nits that would be good to improve, but they aren't blockers to me.

mawinter69 committed:
feedback 2

don't add token when choosing date in the past

directly warn after creation when about to expire

mawinter69 committed:
prettier

janfaracik reviewed (approved):
Thanks for the changes :) LGTM.

timja reviewed (approved):
Re-tested, changes looks good

-----

/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!
